primitives/src/valid_proof_of_investor.rs

```diff
@@ -1,38 +1,27 @@
 use crate::{
-    scalar_blake2_from_bytes, CddId, Claim, Context, IdentityId, InvestorZKProofData, Proposition,
-    Ticker,
+    scalar_blake2_from_bytes, CddId, Claim, IdentityId, InvestorZKProofData, Scope, Ticker,
 };
 use cryptography::claim_proofs::ProofPublicKey;
 use curve25519_dalek::{ristretto::CompressedRistretto, scalar::Scalar};
 
 // ZKProofs claims
 // =========================================================
 
-/// Proposition that checks if any of its internal claims exists in context.
+/// Data structure that checks if any of its internal claims exists in context.
 #[derive(Clone)]
 #[cfg_attr(feature = "std", derive(Debug))]
-pub struct ValidProofOfInvestorProposition {
-    /// The Investor proof should be associated to this ticker.
-    pub ticker: Ticker,
-}
-
-impl Proposition for ValidProofOfInvestorProposition {
-    /// Evaluate proposition against `context`.
-    fn evaluate(&self, context: &Context) -> bool {
-        context
-            .claims
-            .iter()
-            .any(|claim| self.evaluate_claim(claim, context))
-    }
-}
+pub struct ValidProofOfInvestor;
 
-impl ValidProofOfInvestorProposition {
+impl ValidProofOfInvestor {
     /// Evaluates if the claim is a valid proof.
-    fn evaluate_claim(&self, claim: &Claim, context: &Context) -> bool {
+    pub fn evaluate_claim(claim: &Claim, id: &IdentityId) -> bool {
         match claim {
-            Claim::InvestorZKProof(ref _ticker_scope, ref scope_id, ref cdd_id, ref proof) => {
-                let message = InvestorZKProofData::make_message(&context.id, &self.ticker);
-                Self::verify_proof(cdd_id, &context.id, scope_id, &self.ticker, proof, &message)
+            Claim::InvestorZKProof(ref scope, scope_id, cdd_id, proof) => {
```

Comment:
Would it make sense to define the first parameter of `Claim::InvestorZKProof` as a `Ticker` instead of `Scope`? Since only the `Ticker` part is processed, I think that defining it as `Scope` is too general.

Reply:
Ideally it should be possible to use any of the Scope enum variants. @satyamakgec Can we generalise the proof validation to work with any variant?

Reply:
Yes @adamdossa Ideally it is possible to do so but I need `Ticker` explicitly to validate the proof at the time of adding claim. So either I need to add one extra parameter i.e Ticker within the `InvestorZkProof` variant and allow a general`Scope` so I can access the ticker without worrying about the scope (either it is issuer identity/ ticker/asset identity) or restrict the Scope as Ticker only for storage saving and easiness. 

Scope claims are tightly coupled with `Ticker` so those claims can't be re-used for any other assets that's why generalize scope will not help anyway. 

I can replace `Scope` with `Ticker` as @vkomenda suggested but it will lead more confusion because in compliance-manager we are querying the scope claims on the basis of scope as a ticker and if the variant doesn't posses any scope then how we will assume to query based on scope. So to make everything consistent I keep the `Scope` as the first parameter but restrict it for Ticker only.